### ISSUE TITLE
Tabellen - Standardsortierung als Textfeld

### ISSUE THREAD

**Issue by tbaddade** [yakamara/redaxo_yform#99]
Bis jetzt kann man nur eine Spalte auswählen. Optional könnte man ein Textfeld anbieten, so dass man die Sortierung selbst festlegen kann. Bsp.: `country, zip, title`

**Comment by alexplusde** [comment]
Das stimmt. Sobald eine beschränkte `prio` im Spiel ist, wäre das Gold wert: Dann könnte man erst nach Beschränkung und dann nach prio sortieren und so die Einträge korrekt gruppiert und sortiert ausgeben.

**Comment by dergel** [comment]
hmm.. also ganz frei ? sowas will ich eigentlich so viel wie möglich vermeiden.. fehler etc.

**Comment by alexplusde** [comment]
etwas Offtopic:

An mehreren Stellen in YForm und Redaxo würde sich dafür JQuery Chosen anbieten. Das würde die freie Eingabe von Feldern, aber dennoch datenbankgestützt, extrem erleichtern.

![image](https://cloud.githubusercontent.com/assets/3855487/17548323/ab020c96-5eeb-11e6-975a-fd576f057f31.png)

![image](https://cloud.githubusercontent.com/assets/3855487/17548447/3d32e950-5eec-11e6-8916-7d6fe33fc69b.png)


* Auswahl von Feldnamen in den Tabelleneigenschaften, bei be_relation und diversen anderen Values / Validates
* Darstellung von kommagetrennten Werten (wie bspw. bei der Eingabe von Dateiendungen) in Form von Tags
* Durchsuchbare Dropdowns, bspw. beim Prio-Feld. Manchmal habe ich Formulare mit 30 Feldern und muss lange suchen, wo sich das richtige "Nach $feld" befindet.

**Comment by skerbis** [comment]
oder select2 https://select2.github.io/examples.html

**Comment by dergel** [comment]
und order by müsste auch irgendwie rein .. die varianten sind eigentlich ganz cool (ich glaube aber, dass  die reihenfolge nicht üübernommen wird, oder) ..

**Comment by alexplusde** [comment]
Stimmt, das mit der Reihenfolge könnte ein Problem werden... https://github.com/harvesthq/chosen/pull/197

**Comment by alexplusde** [comment]
Für ASC/DESC könnte man ja auch (notfalls) die Werte doppelt im select2 oder chosen hinterlegen, also [field1 ASC] [field1 DESC] und mit den richtigen Namen ausstatten 
[Feldname1 aufsteigened][Feldname1 absteigend]

**Comment by alexplusde** [comment]
Das Feature wäre für yform_usability relevant,  wenn man beschränkte Prio-Felder hat.

**Comment by gharlan** [comment]
Wenn du einfach nur nach Prio sortierst, passt es nicht?

**Comment by alexplusde** [comment]
![image](https://cloud.githubusercontent.com/assets/3855487/23767658/a11636ca-0509-11e7-8233-edf3611e7576.png)

Felddefinition:
![image](https://cloud.githubusercontent.com/assets/3855487/23767678/be1d8674-0509-11e7-844f-7944ecdec433.png)

Nein, die zusätzliche Beschränkung wird übergangen.

**Comment by gharlan** [comment]
Wenn ich ein beschränktes Prio-Feld benutzt habe, dann bisher immer nur in Tabellen, die nur über ein "multiple 1-n"-Feld erreichbar waren.
Also habe die Tabellen in der Navigation immer ausgeblendet.
Also konnte man bei mir die Tabelle auch nur beschränkt pro Kategorie öffnen, also immer nur ein Prio-Bereich.

**Comment by gharlan** [comment]
Bei dir sind die Prios übrigens vor allem auch falsch. Denn sie sind ja gar nicht pro Kategorie aufsteigend, sondern Kategorieübergreifend.
Ist also nicht nur ein Anzeigeproblem (Sortierung).

**Comment by alexplusde** [comment]
Das ist yform_usability geschuldet, weil ich in dem Screenshot noch Datensätze sortiert habe, um zu schauen, ob das irgenwie greift.

**Comment by alexplusde** [comment]
> Also konnte man bei mir die Tabelle auch nur beschränkt pro Kategorie öffnen, also immer nur ein Prio-Bereich.

Wie machst du das?

**Comment by gharlan** [comment]
Bei der Tabelle auswählen "In Navigation verstecken" (oder so ähnlich).
Dann in der Kategorie ein "multiple 1-n"-Feld anlegen (falls nicht schon vorhanden), dort "In Liste verstecken" NICHT anhaken.
Dann sollte in der rex_list zur Kategorie-Tabelle in jeder Zeile ein Link zu den zugehörigen Datensätzen sein.

**Comment by gharlan** [comment]
Übrigens, das Grund-Issue hier unterstütze ich. Fände also eine flexiblere Sortierauswahl auch toll, wie auch immer man es am besten löst.

**Comment by alexplusde** [comment]
Sorry, ich versteh's noch nicht. Welche Tabelle meinst du bei "Tabelle"? Die Tabelle mit den Artikeln sollte nicht in der Navigation versteckt werden, sondern bearbeitbar bleiben. be_relation-Felder lassen sich auch nicht sortieren, wie soll das mit der Prio funktionieren?

**Comment by gharlan** [comment]
Vorweg: Es kann schon sein, dass es für dich, in deinem Fall nicht passt. Ich meinte ja nur, bei meinen bisherigen Fällen, wo ich beschränkte Prio-Felder genutzt habe, hatte ich sowieso so eine Struktur, wo sich die entsprechende Tabelle nur mit Beschränkung betrachten ließ.

Habe mal auf die Schnelle einen solchen Fall konstruiert, um es anhand von Screenshot zu zeigen.
Habe zwei Tabellen, angelegt, Produkte und Produkt-Kategorien. 
Die Produkt-Tabelle lasse ich in der Navigation verstecken. Die Navi sieht also so aus:

<img width="262" alt="screenshot 2017-03-11 15 24 07" src="https://cloud.githubusercontent.com/assets/330436/23824042/aa4b720c-066f-11e7-8284-3fb609485720.png">

In den Produkten gibt es ein Relationsfeld `category_id` (select single). In den Kategorien gibt es eine Rückreferenz darauf über ein "Popup (multiple 1-n)"-Feld. Dieses lasse ich in der Liste NICHT verstecken.

Wenn man also über die Navigation die Kategorien öffnet, erhält man das:

<img width="940" alt="screenshot 2017-03-11 15 27 21" src="https://cloud.githubusercontent.com/assets/330436/23824054/e657b38c-066f-11e7-8d87-49b30cef308b.png">

Bei jeder Kategorie gibt es den Link "Produkte". Über diesen Link kommt man dann zur Produkttabelle, aber beschränkt auf diese eine Kategorie:

<img width="945" alt="screenshot 2017-03-11 15 33 19" src="https://cloud.githubusercontent.com/assets/330436/23824063/1550daa6-0670-11e7-962c-aae4668fbb31.png">

Hätte ich jetzt ein beschränktes Prio-Feld (beschränkt auf die Kategorie), wäre also in jeder Ansicht der Produkttabelle immer nur ein solcher Teilbereich der Prios zu sehen.
Somit passte es bei mir dann auch, wenn ich die Tabelle einfach nur nach Prio sortieren lasse.

Klar, der Nachteil ist, dass es umständlicher ist, zur Produkttabelle zu gelangen. Aber wie gesagt, in den Fällen, wo ich bisher beschränkte Priofelder genutzt habe, wollte ich das sowieso so haben, dass die jeweilige Tabelle nicht als ganzes erscheint, sondern immer nur für den Teilbereich.

**Comment by alexplusde** [comment]
Ach so geht das! Ich glaube, das ist ein Feature, das keiner kennt außerhalb von Yakamara.

**Comment by gharlan** [comment]
Ein paar müssten es schon kennen, habe es jedenfalls schon häufiger erklärt, aber waren eigentlich immer Einzelgespräche. Daher kann es schon sein, dass es der Masse unbekannt ist. ;)

Eingeführt wurde es in XForm 4.10: https://github.com/yakamara/redaxo_xform/commit/6879d61e07f84f4ca36c978d8f89e1396e8ecb6a

**Comment by alexplusde** [comment]
Nichtsdestotrotz habe ich gerade den Fall, bei dem ich erst nach Datum sortieren möchte (Termine), und dann nach prio (Termine auf Anfrage) - es wäre klasse, wenn man hier zumindest bis zu 3 Sortierungen angeben kann oder optional eine andere Lösung findet wie oben angedacht...

**Comment by tbaddade** [comment]
Ich wäre für das freie Textfeld, da kann man auch direkt ASC und DESC angeben.

> sowas will ich eigentlich so viel wie möglich vermeiden.. fehler etc.

Dann müsste man das bei den Values auch nachziehen, sodass dort eben keine eigenen SQL-Statements mehr hinterlegt werde können. Wobei ich die freie Eingabe eher als Vorteil sehe und vermutlich auch nur von Devs genutzt werden.

**Comment by alexplusde** [comment]
@tbaddade du hast recht. Man müsste auch abfangen, ob es bspw. Felder nicht mehr gibt, wenn sie gelöscht wurden. Freitext ist sicher am besten.
